Example context

juanignaciosl opened the pull request:
Auth API is going to be released really soon, and this PR is blocked because of that.

I've been doing some tests with the [`auth-api` branch of SQL API](https://github.com/CartoDB/CartoDB-SQL-API/tree/auth-api), which is still under heavy development.

Assuming that Auth API fits the problem of having a managed user with example data, we should use the (new) default public key token, `default_public`. That would grant automatically read access to every public dataset. I think that this is the simplest approach, but we should come up with a proper way of checking if it's a valid authentication. Currently, this is done:

https://github.com/CartoDB/cartoframes/blob/777ff514d7f1930c0a9af9e2152b3d6ee5e33702/cartoframes/context.py#L117-L121

Nevertheless, as we've said, that's a hack. I've requested a proper fix here: CartoDB/cartodb#13556


I'd keep this PR on hold until Auth API is released (it's going to happen in a few days). Once it's released, we can change `_is_authenticated ` so it uses a valid approach with Auth API and release this, with `cartoframes` user. What do you think, @andy-esch ?

_EDIT: also related: CartoDB/carto-python/issues/69_

andy-esch commented:
Nice :) Exciting to be moving to the new Auth API. Since we're using the [`APIKeyAuthClient`](https://github.com/CartoDB/carto-python/blob/master/carto/auth.py#L29) client from the carto-python sdk, what would it mean for updating the rest of the cartoframes code?

As far as functionality, the only thing that we're missing is an easy way to explore what datasets are in the account. I proposed a `list` method in https://github.com/CartoDB/cartoframes/issues/170, but if we go the way of creating a special `Example` CartoContext, then that method would have to be in the CartoContext class as well -- which I think would be a good thing. We could also publish a list of available datasets in the docs and not create a new `list` method, but the docs would need to be kept in sync independently of the cartoframes.carto.com account, which wouldn't be ideal.

We thought of adding more dataset managment functionality in https://github.com/CartoDB/cartoframes/issues/353 based on some user feedback.

andy-esch commented on cartoframes/examples.py:
```diff
@@ -0,0 +1,5 @@
+from cartoframes import CartoContext
+
+def exampleContext():
```
How about just `Examples` like you proposed in https://github.com/CartoDB/cartoframes/issues/170#issuecomment-364934921? I feel the user doesn't need to specifically know this is another `CartoContext` (which for us is an implementation detail).

juanignaciosl replied:
Not sure. If the user doesn't know that it's a `CartoContext`, how can he or she know what API can be used? I don't think that documentation should duplicate the `CartoContext` one.

Whatever the question is, uppercasing the first letter is pythonic?

Both questions are driven by my Python ignorance, you know 😅 I'm asking just to understand the best way to do this, I'll do what you think is better :-)

andy-esch commented on cartoframes/examples.py:
```diff
@@ -0,0 +1,5 @@
+
+def exampleContext():
+    """Returns a CartoContext corresponding to a CARTO account with example data"""
+    return CartoContext(base_url='https://cartoframes.carto.com', api_key='default_public')
```
Could you limit this line and the docstring above it to 79 lines please?

andy-esch replied:
I use both `pep8` and `flake8` for linting purposes

juanignaciosl replied:
Ok! I just enabled pep8 warnings at IntelliJ (I'm trying to use it as swiss-army knife, with language plugins), and [added flask8](https://foxmask.trigger-happy.eu/post/2016/02/17/pycharm-running-flake8/) so it complains:

```
/usr/local/bin/flake8 --max-complexity 10 /Users/juanignaciosl/Development/carto/cartoframes/cartoframes/examples.py
/Users/juanignaciosl/Development/carto/cartoframes/cartoframes/examples.py:5:80: E501 line too long (83 > 79 characters)
/Users/juanignaciosl/Development/carto/cartoframes/cartoframes/examples.py:6:80: E501 line too long (91 > 79 characters)

Process finished with exit code 1
```

andy-esch commented on cartoframes/examples.py:
```diff
@@ -0,0 +1,5 @@
+from cartoframes import CartoContext
+
+def exampleContext():
+    """Returns a CartoContext corresponding to a CARTO account with example data"""
```
We're going to need to write a full docstring on the usage of this

juanignaciosl replied:
I guess that this depends on hiding that it's a `CartoContext` or not. As I was proposing exposing it, it might not need documentation. What documentation were you missing?

Uhm, `CartoContext` hasn't any method for dataset discoverability within the account, has it? Maybe, instead of documenting the available datasets we could add listing of available datasets.

andy-esch commented on cartoframes/examples.py:
```diff
@@ -0,0 +1,5 @@
+from cartoframes import CartoContext
```
We should add a simple module docstring here for documentation purposes.

juanignaciosl replied:
Ok. I'll wait for the decision on the approach for `exampleContext`.

andy-esch commented:
@juanignaciosl, how do you suggest testing this? Since `CartoContext` has tests already, we'd be testing whether datasets can be downloaded by this mechanism?

juanignaciosl commented:
Answering some of your questions, @andy-esch :

- About testing: maybe we could check the existence of a well-known dataset, and the impossibility of editing it (so we don't make this example writable by mistake, either from CARTO Frames or because of the platform). I've added a test that

> Since we're using the APIKeyAuthClient client from the carto-python sdk, what would it mean for updating the rest of the cartoframes code?

We're far from that yet. Auth API will only be released with support for Maps and SQL APIs, and only partially, unofficially for itself (for example, Auth API key listing). I wouldn't consider a full migration yet, but we can begin adopting it for new usages or new needs.

> I proposed a list method in #170, but if we go the way of creating a special Example CartoContext, then that method would have to be in the CartoContext class as well -- which I think would be a good thing.

I've added a `datasets` method at `CartoContext`. It returns a `carto-python` list of `Dataset`s, I think that there's no point on wrapping it.

> How about just Examples like you proposed in #170 (comment)

As you see, I've gone back to a `examples` factory method instead of a class. What do you think? In my opinion, returning a `CartoContext` and improving it (with methods like `datasets`) is better, but it's true that we're exposing that it's a `CartoContext`. If we actually wanted to hide this, we should create a `Examples` wrapper and expose the desired interface with delegation methods, such as a new `datasets` and so on. What do you think?

juanignaciosl commented:
@andy-esch , [tests are passing](https://travis-ci.org/CartoDB/cartoframes/jobs/347244862) 🙌 I've marked this as blocked because this can't be released until Auth API release. Anyway, I think that it's ready for a CR and continue the discussion about what API should we expose (specific for examples? CartoContext?). Take a look when you can

andy-esch commented on cartoframes/context.py:
```diff
@@ -167,6 +166,9 @@ def read(self, table_name, limit=None, index='cartodb_id',
         return self.query(query, decode_geom=decode_geom)
 
+    def datasets(self):
+        return DatasetManager(self.auth_client).all()
```
Nice :) Since this returns a list of `Dataset` instances, we should probably have it be a list of names instead.

```python
return [d.name for d in DatasetManager(...).all()]
```

Another thing is that `DatasetManager().all()` is slow for my account (~1250 tables). Is there another way to list tables that's faster? My guess is that we should avoid querying the information schema table, but this would probably be faster.

juanignaciosl replied:
Ok. Indirectly, you've just answered my biggest doubt: I wasn't sure if a CARTO Frames function (`datasets`) should return instances of a class that is defined in CARTO Python SDK. I'll take a look at it because only the name is going to be returned, we should reduce the retrieved information a lot (therefore making it faster).

andy-esch replied:
Yeah, good point about carto python sdk `Dataset` object. One thing that we can think about for the near future planning is how this can look for the cartoframes Analysis API's data source objects: http://cartoframes.readthedocs.io/en/prototype-table-objects/cartoframes.analysis.html#data-sources

Should the returned datasets be a to-be-created `Table` instance? Or should we hold off for now, and just have it as a string for the table name, and then update it to `Table` later once we write the spec? I support this second option.

andy-esch commented on cartoframes/context.py:
```diff
@@ -105,25 +105,24 @@ def __init__(self, base_url=None, api_key=None, creds=None, session=None,
-            self.sql_client.send(
-                'select * from information_schema.tables limit 0')
-        except CartoException as err:
+        if not auth_api_client.is_valid_api_key():
```
Why pass the auth api client as an argument instead of doing this:
```python
if not self.auth_api_client.is_valid_api_key():
    ...
```

juanignaciosl replied:
Oh, I tend to pass as parameters whatever is needed, but this is certainly an abuse of that style. I'll change it.

andy-esch commented:
@juanignaciosl I left a couple of comments -- I love that this is reducing the amount of code for the auth piece :)

About the examples API, I think I'm in favor of providing a few functions that use the example context. We could still let the users know about the `example_context`, which they could use for some basic functionality (like mapping datasets before reading them into a dataframe). That reminds me that I want better errors for things like data obs usage (which I think says something about quota which the user might not understand) and other limits.

This is how I envision the API working:
> examples.py
```python
example_context = CartoContext(...)

def load_taxi():
    return example_context.read('taxi_dataset')

def load_mexico_city():
    return example_context.read('mexico_city_metro')
```

Usage in a script would be like:
```python
from cartoframes import examples

# get dataframe of taxi data
taxi_df = examples.load_taxi()
```
I can't see having more than 5-10 of these `load_x` functions, so maintenance is not hard. It also closely mirrors what scikit-learn does for their example datasets: http://scikit-learn.org/stable/modules/classes.html#module-sklearn.datasets

Another question is whether `load` is the correct word here when the data is actually `fetched`, but that's not too important maybe.

What do you think?

juanignaciosl commented:
@andy-esch I had also thought about creating some methods that would allow users to discover data even without listing tables, but I dismissed it because coupling functions to data felt weird. Anyway, if that's common in this context, let's do it. What do you think about creating a `Examples(CartoContext)` class with current `examples` credentials, and adding `load_X` functions there?

PS: I don't have strong feelings about `fetch` vs `load`. You're right, `fetch` is more accurate, but that's often just a convention, and depending on it, `load` might be more convenient. Up to you.

juanignaciosl commented:
@andy-esch take a look. Performance of `datasets` (now `tables`) should be ~5x better.

I've finally opted by going to `Table` and hiding `Dataset`, in order to avoid coupling. I've created a factory based on `Dataset` so we could easily aggregate the source dataset inside `Table` and delegate some functionality to it. I'm not too happy with this approach, though. I've talked to @danicarrion about this and I must admit that once I begun coding I didn't come up with a approach that truly convinced me. What do you think, folks?

danicarrion commented:
I guess my question would be, how do we think this `Table` class is going to evolve? What do you expect it to become in the future?

andy-esch commented:
@danicarrion, we see it being one of a couple of node types in the cartoframes Analysis API: http://cartoframes.readthedocs.io/en/prototype-table-objects/cartoframes.analysis.html 
`Table` would inherit from `Query`, which is detailed here: http://cartoframes.readthedocs.io/en/prototype-table-objects/cartoframes.analysis.html#cartoframes.analysis.Query

juanignaciosl commented:
@andy-esch uhm, I hadn't thought about `Table` that way at all. I was thinking more about metadata (name, schema...), not about the data itself.

I would've said that "executing a `Query` _returns_ a `Table`", but I'm not sure if we can say that a `Table` *_is_* a `Query`, so I'm not sure if `Table` can't inherit from `Query`.

Anyway, I've been trying to come up with a case that actually shows that a `Table` _is not_ a `Query`, but I haven't succeeded, so maybe I'm wrong.

@danicarrion does this look right to you?

@andy-esch anything else that you'd improve or change?

The scope of this PR has grown since we begun, but as it's still blocked by some Auth API deployments I think that beginning this discussion around Analysis API modeling is really nice :-)

andy-esch commented on cartoframes/analysis.py:
```diff
@@ -0,0 +1,8 @@
+class Table:
```
💥 

How about adding a
```python
def __repr__(self):
    return 'Table(name={})'.format(self.name)
```
so that the Table objects print nicely?

andy-esch commented on cartoframes/context.py:
```diff
@@ -167,6 +167,19 @@ def read(self, table_name, limit=None, index='cartodb_id',
+            show_uses_builder_features='false',
+            show_synchronization='false',
+            load_totals='false')
+        return (Table.from_dataset(d) for d in datasets)
```
Why did you opt for a generator here? I'm guessing that the most common use case for this method is to get a full list of all the tables and it might not be obvious that this should be iterated over.

juanignaciosl replied:
When languages provide lazy declaration I use it by default, and other people have recently suggested using iterators instead in other similar cases. Anyway, switching to list if you think it's better 👍

andy-esch commented:
I think we're in agreement about `Table` :) The way I see it `Table` would be just an abstract representation of a table in CARTO. Like you said, it'd just have metadata that would allow other operations to act on it and perform database operations in CARTO's cloud.

For me `Table` is just a convenience for labeling a table in the database instead of having to do `SELECT * FROM table` -- and allows an easy interface for users to create nodes from all rows and columns of a table.

When I say `Table` should inherit from query, I mean that `Table` is just `SELECT * FROM table`, so any operations that you perform on `Table` should be perform-able on `Query` too. Just like Builder analysis, the node doesn't have to be the original table query, but can be an arbitrary query.

Happy to have this discussion :)

juanignaciosl commented:
> When I say Table should inherit from query, I mean that Table is just SELECT * FROM table, so any operations that you perform on Table should be perform-able on Query too. Just like Builder analysis, the node doesn't have to be the original table query but can be an arbitrary query.

In fact, I've suffered a lot with that convention in the past. There are parts of the backend that make that kind of assumption, and it's really confusing. For example, if you allow it to be an arbitrary query, you could do a join to another table, so your `Table` node is actually querying two tables, and some mismatching arises. As currently there's no need for that inheritance, I'd suggest to keep `Table` as simple as possible and reconsider inheritance later. What do you think?

andy-esch commented:
I see no need to introduce anything about the `Query` class here, so the current implementation of `Table` looks 👌

andy-esch commented:
... and I'm really glad you have lessons learned about how the handling of data nodes like this!

juanignaciosl commented:
Cool, I think there's an agreement here :-)

API keys is still not released, so we're still blocked. Anyway, what are your thoughts about remaining stuff, @andy-esch ? Probably, more documentation, and then add the test datasets, right?

andy-esch commented on cartoframes/examples.py:
```diff
@@ -0,0 +1,17 @@
+# EXAMPLE_API_KEY = 'default_public'
+
+
+class Examples(CartoContext):
```
I'm in the class camp now because we can overwrite the CartoContext methods that we don't want to expose 👍 

Something like the following for `write`, `data*`, and `delete`:

```python
def data(self):
    raise RuntimeError('CartoContext.data disabled for the Examples CartoContext')
```

andy-esch commented:
Looking great :) I wonder if we can go even further in simplifying it for the user. These are three ways I imagine using it.

```python
from cartoframes import Examples

# 1. current implementation
df = Examples().read_taxi()

# 2. desired additional usage - loaders are functions in the module
# this feels more natural to me / requires less of the user
from cartoframes import examples
df = examples.read_taxi()

# or
from cartoframes.examples import read_taxi
df = read_tax()

# 3. super users
# this is currently possible
example_context = Examples()

# but it would be easier if the user didn't have to instantiate it
# if they want to use the special cartocontext, they can just import it
from cartoframes.examples import example_context

df = example_context.query('''
    SELECT s.*, count(t.*)
    FROM taxi as t
    JOIN ny_census_tract as s
    ON st_intersects(t.the_geom, s.the_geom)
    ''')
```

The first method is nice and easy to do with the current implementation. To me, having to create the class instance seems like a step that we don't necessarily need the user to go through. To support the second and third options, we could offer a pre-made context like this in `examples.py`:

```python
"""examples.py"""

example_context = Examples()

# and define the loading methods as functions instead
def load_taxi(limit=None):
    return example_context.read('taxi_dataset', limit=limit)
```

And we could re-use these functions in the class:

```python

class Examples(CartoContext):
    def __init__():
        pass

    def read_taxi(self, limit=None):
        return load_taxi(limit=limit)
```

juanignaciosl commented:
From 2., I've chosen the second alternative, because the first one is closer to the 1. implementation.

For `example_context` I've implemented it the other way around: `Examples.read_taxi` is the `read` call, and `load_taxi` loader function is a call to `example_context.read_taxi`. Otherwise, because of the order of the definitions, the compiler would complain. That's likely to be something that I'm missing from Python, but the resulting API is the same :-)

What else? :-) I've added the "blocked" label back because Auth API is unreleased yet, targeted for CARTO Locations, but I'd love to have this ready for that date!

andy-esch commented:
🎉 Excited to experiment with this a bit! I'll ping back on here once I explore it a bit. I'll add a couple of datasets to the cartoframes account as well.

andy-esch commented:
Everything is working like a charm when I try with accounts that have the Auth API enabled :)

The tasks I still see for this PR:

1. New datasets and read functions: We're working on fleshing out the cartoframes docs and building some guides and examples, which will have some dataset dependencies. I'll collect all of those and list them in this ticket as new `read..` functions to that need to be added.
1. Documentation for example methods (I'm happy to write this if you don't want to)

There are some other tasks that I [listed here](https://github.com/CartoDB/cartoframes/issues/412) that need to be completed before we merge this in. I'm going to start tackling those.

andy-esch commented:
Now that Auth API is released _and_ the [developer docs](https://carto.com/developers/cartoframes/guides/Quickstart/) are in place, I'll start making a list of the function names we need for the files this week or early next week.

andy-esch commented:
@juanignaciosl I added the functions for the new datasets, as well as some documentation around the functions. 

One thing I noticed with the Auth API is that the postgres search path has `"$user"` as the first element when authenticating with `default_public`. This has an effect on the `CartoContext._is_org_user()` method -- which once master was merged in falsely implies that the user is in an org. This change was introduced in https://github.com/CartoDB/cartoframes/pull/417

juanignaciosl commented:
I've opened CartoDB/CartoDB-SQL-API/issues/483, because that looks like a bug.

Anyway, coupling "being an org user" to `search_path` is fragile. Assuming that it's a bug, what do you think about removing `is_org` and replacing `_is_org_user` with `_default_schema`, returning the first `search_path` entry?

andy-esch commented:
Replacing with schema name works for me. There are places where there are Boolean checks on this elsewhere so it will require a bit of refactoring

andy-esch commented:
@jgoizueta mentioned that another option is to use an as-yet-to-be-created function that returns what we want:

```

cartodb_user_825452d9-9344-4ca7-a0e4-6939766806cc_db=# select cdb_conf_getconf(‘user_config’);
                 cdb_conf_getconf
---------------------------------------------------
 {“is_organization”: false, “entity_name”: “jgoi”}
(1 row)
```

juanignaciosl commented:
While we could find workarounds at SQL API level, I think that the (Rails)
REST API is more appropriate, because it's user metadata. Nevertheless, I
couldn't find any public, unauthenticated endpoint that were valid. IMHO,
if we eventually need this, we could add `organization_name` to `/me`
endpoint, and display it publicly.

On Wed, May 2, 2018 at 2:07 PM, Andy Eschbacher <notifications@github.com>
wrote:

> @jgoizueta <https://github.com/jgoizueta> mentioned that another option
> is to use an as-yet-to-be-created function that returns what we want:
>
>
> cartodb_user_825452d9-9344-4ca7-a0e4-6939766806cc_db=# select cdb_conf_getconf(‘user_config’);
>                  cdb_conf_getconf
> ---------------------------------------------------
>  {“is_organization”: false, “entity_name”: “jgoi”}
> (1 row)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/CartoDB/cartoframes/pull/382#issuecomment-385955537>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA48aMFil_jd-kjaQn2zL6hT8s7XRvURks5tuaFzgaJpZM4SHN8J>
> .
>



-- 
*Juan Ignacio Sánchez Lara*
CARTO Backend Tech Lead
Twitter: @juanignaciosl <https://twitter.com/juanignaciosl>

*CARTO <http://www.carto.com> — Predict Through Location*

Live sessions to learn CARTO <http://carto.com/webinars/>

juanignaciosl commented:
@andy-esch I postponed the discussion about the convenience of removing `_is_org_user` because I'd like to stop increasing the scope of this :D. Instead, I fixed the known issue by switching from `show search_path` to `select current_schemas('f')`. I think that it should be enough to unlock (and maybe release?) this. Let me know!

andy-esch commented:
100% agree about limiting the scope. This has creeped quite a bit! Agree we should fix the is org user in a separate PR. I'll do some more ad hoc testing on my machine and report back next week (at a conference giving a talk today).

michellemho commented on cartoframes/examples.py:
```diff
@@ -0,0 +1,358 @@
+    .. code::
+
+        from cartoframes.examples import read_taxi
+        df = read_taxi()
```
Change to `read_nat`

juanignaciosl commented on cartoframes/examples.py:
```diff
@@ -0,0 +1,381 @@
+"""Download, preview, and query example datasets for use in cartoframes
```
@andy-esch I love your point about not even needing a CARTO account to try it, I'd write it down here, and maybe in the [landing features list](http://cartoframes.readthedocs.io/en/latest/#features) with a link.

andy-esch replied:
Great suggestion, added to both :)

juanignaciosl commented on cartoframes/examples.py:
```diff
@@ -0,0 +1,381 @@
+:py:class:`CartoContext <cartoframes.context.CartoContext>` that is
+authenticated against all public datasets in the https://cartoframes.carto.com
+account. This means that besides reading the datasets from CARTO, users can
+also create maps from these datasets.
```
This will become a problem because of the limit of named maps. I'd explain that "maps created at this account will be periodically removed", and schedule a cleanup.

juanignaciosl commented on cartoframes/examples.py:
```diff
@@ -0,0 +1,381 @@
+
+    .. code::
+
+        from cartoframes.examples import example_context
```
N00b people like me love those examples, copy-and-paste friendly, that include `import` lines in every snippet :heart:
